homeassistant/components/modbus/__init__.py

```diff
@@ -206,6 +211,8 @@ def connect(self):
         """Connect client."""
         with self._lock:
             self._client.connect()
+            if self._delay > 0:
```

Comment:
Please adjust the timeout instead of sleeping

see ` pymodbus/client/sync.py `
```
# --------------------------------------------------------------------------- #
# Modbus TCP Client Transport Implementation
# --------------------------------------------------------------------------- #
class ModbusTcpClient(BaseModbusClient):
    """ Implementation of a modbus tcp client
    """

    def __init__(self, host='127.0.0.1', port=Defaults.Port,
        framer=ModbusSocketFramer, **kwargs):
        """ Initialize a client instance

        :param host: The host to connect to (default 127.0.0.1)
        :param port: The modbus port to connect to (default 502)
        :param source_address: The source address tuple to bind to (default ('', 0))
        :param timeout: The timeout to use for this socket (default Defaults.Timeout)
        :param framer: The modbus framer to use (default ModbusSocketFramer)

        .. note:: The host argument will accept ipv4 and ipv6 hosts
        """
        self.host = host
        self.port = port
        self.source_address = kwargs.get('source_address', ('', 0))
        self.socket = None
        self.timeout = kwargs.get('timeout',  Defaults.Timeout)
        BaseModbusClient.__init__(self, framer(ClientDecoder(), self), **kwargs)
```

Reply:
If the timeout doesn't fix it you might need to do a select with a timeout on the socket to verify its ready

See _recv in ModbusTcpClient
```
            ready = select.select([self.socket], [], [], end - time_)
```

Reply:
@bgianfo Adjusting timeout doesn't work for this device (Huawei SUN2000). You need to wait between connecting and sending (or try three times in a row).

Reply:
I would love to avoid the sleep, but the timeout parameter works on the socket before the connect, and as @Emilv2 noted the delay is after a succesfull connect() and the first send().

If there is a function in core like wait(secs) please tell me and I will implement that.

Reply:
This can't be async with a sync client.  This would need to be converted to async

```
from pymodbus.client.sync import ModbusSerialClient, ModbusTcpClient, ModbusUdpClient
```

pymodbus has an async client.
https://pymodbus.readthedocs.io/en/latest/source/library/pymodbus.client.html